pallets/rbac/src/functions.rs

```diff
@@ -1,434 +1,514 @@
+		ensure!(Self::has_unique_elements(permissions.clone()), Error::<T>::DuplicatePermission);
+		let pallet_id_enum = pallet.to_id_enum();
+		let pallet_id = pallet_id_enum.to_id();
+		Self::is_role_linked_to_pallet(pallet_id_enum, &role_id)?;
+
+		let role_permissions = <PermissionsByRole<T>>::get(&pallet_id, role_id);
+		for id in permissions.clone() {
+			ensure!(!role_permissions.contains(&id), Error::<T>::PermissionAlreadyLinkedToRole);
+		}
+		<PermissionsByRole<T>>::try_mutate(pallet_id, role_id, |role_permissions| {
+			role_permissions.try_extend(permissions.into_iter())
+		})
+		.map_err(|_| Error::<T>::ExceedMaxPermissionsPerRole)?;
+		Ok(())
+	}
+
+	/*---- Helper functions ----*/
+
+	/// Authorization function
+	///
+	/// Checks if the user has a role that includes the specified permission.
+	/// ### Parameters:
+	/// - `user`: The account to validate.
+	/// - `pallet_id`: The unique pallet identifier.
+	/// - `scope_id`: The scope context in which the permission will be validated.
+	/// - `permission_id`: The permission the user must have.
+	fn is_authorized(
+		user: T::AccountId,
+		pallet: IdOrVec,
+		scope_id: &ScopeId,
+		permission_id: &PermissionId,
+	) -> DispatchResult {
+		let pallet_id_enum = pallet.to_id_enum();
+		let pallet_id = pallet_id_enum.to_id();
+		Self::scope_exists(pallet_id_enum.clone(), scope_id)?;
+		Self::permission_exists(pallet_id_enum, permission_id)?;
+		// get roles the user has in this scope
+		let user_roles = <RolesByUser<T>>::get((user, pallet_id, scope_id));
+		// determine if one of the roles has the requested permission
+		let has_permission = user_roles
+			.iter()
+			.any(|r_id| <PermissionsByRole<T>>::get(pallet_id, r_id).contains(permission_id));
+		ensure!(has_permission, Error::<T>::NotAuthorized);
+		Ok(())
+	}
+
+	/// User role validation function
+	///
+	/// Checks if the user has at least one of the specified roles.
+	/// ### Parameters:
+	/// - `user`: The account to validate.
+	/// - `pallet_id`: The unique pallet identifier.
+	/// - `scope_id`: The scope context in which the permission will be validated.
+	/// - `role_ids`: A list of roles to validate.
+	fn has_role(
+		user: T::AccountId,
+		pallet: IdOrVec,
+		scope_id: &ScopeId,
+		role_ids: Vec<RoleId>,
+	) -> DispatchResult {
+		let pallet_id_enum = pallet.to_id_enum();
+		Self::scope_exists(pallet_id_enum.clone(), scope_id)?;
+		let user_roles = <RolesByUser<T>>::get((user, pallet_id_enum.to_id(), scope_id));
+		ensure!(user_roles.iter().any(|r| role_ids.contains(r)), Error::<T>::NotAuthorized);
+		Ok(())
+	}
+
+	/// User any-role validation function
+	///
+	/// Checks if the user has at least one role on the given scope.
+	/// ### Parameters:
+	/// - `user`: The account to validate.
+	/// - `pallet_id`: The unique pallet identifier.
+	/// - `scope_id`: The scope context in which the permission will be validated.
+	fn has_any_role(account: T::AccountId, pallet: IdOrVec, scope_id: &ScopeId) -> bool {
```

Comment:
Due to the needing to "ensure if a user is part of the marketplace," we could use the following code:
```rust 
    fn does_user_have_role_is_this_scope(user: T::AccountId, scope_id: &ScopeId, pallet: IdOrVec) -> bool { 
        let pallet_id = pallet.to_id();
        //option 1
        let rr = UsersByScope::<T>::iter_prefix_values((pallet_id, scope_id)).collect::<Vec<_>>();
        rr.iter().any(|r| r.contains(&user));

        //option 1.1
        let rr2 = UsersByScope::<T>::iter_prefix((pallet_id, scope_id)).collect::<Vec<_>>();
        rr2.iter().any(|r| r.1.contains(&user));

        //option 1.1 compressed
        UsersByScope::<T>::iter_prefix((pallet_id, scope_id)).any(|(role_id, users)| {
            users.contains(&user)
        })
    }
```
Since many roles could be part of a frunique's collection, any of my above options is a better implementation for this problem and future needs. 
Your current implementation could lead to emptiness problems, for example:
When a new entry is created in the RolesByUser storage map, the value of that entry is a vector of roles. But if we remove all the roles from a selected user, the initial entry will still exist with zero values. The contains_key method only checks if the entry exists, and doesn't care about if the value is null or something. 
Use any of my suggested methods and CREATE unit tests for it.